Real-time Collaboration: Add sync connection error handling

shekharnwagh opened the pull request:
## What?
Add sync connection error handling

Creates a provider-agnostic system for sync providers to report connection errors and display appropriate UI to users.

## Why?
Sync providers currently have no standard way to communicate connection errors to users. When a sync provider disconnects or authentication fails, users are left in an unclear state with no feedback about what happened or how to save their work.

This PR provides a centralized system for any sync provider to report connection status and automatically display a modal to users when disconnections occur.

## How?
This implementation is based on the modal from the VIP Real-Time Collaboration plugin:
https://github.com/Automattic/vip-real-time-collaboration/blob/trunk/src/components/post-locked-modal.tsx

- Added a new Redux store (`core/sync-connection`) in `@wordpress/sync` package that tracks connection status for all sync providers
- Created `SyncConnectionModal` component in `@wordpress/block-editor` that displays when any provider reports a disconnection
- Providers can call `setConnectionStatus()` with standard error types (auth failed, too many connections, etc.) or custom messages
- Modal provides users with options to copy their content and safely exit the editor

chriszarate commented:
> But if we can think of cases where an error might be entity specific, we can look into options on how to handle this.

I'm thinking mostly of connection errors that only affect 1 entity because of an n + 1 problem like a connection limit.

shekharnwagh commented:
> > But if we can think of cases where an error might be entity specific, we can look into options on how to handle this.
> 
> I'm thinking mostly of connection errors that only affect 1 entity because of an n + 1 problem like a connection limit.

I'm inclined to show the modal even if a single entity encounters connection issues. This would ensure atomic behavior, where the sync either completes entirely or not at all, unless the issue is specific to a particular entity and due to sync server like auth, connection limit, etc.

chriszarate commented on packages/editor/src/components/sync-connection-modal/index.js:
```diff
@@ -0,0 +1,93 @@
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { error as errorIcon } from '@wordpress/icons';
+import { getConnectionStatusMessage } from '@wordpress/sync';
```
It would be nice not to import the sync package at all and continue to rely purely on `core-data`.

Is there a better pattern than translating codes to user-ready text? I see from `core-data/src/selectors.js` that there are many examples of `Error` objects in state. Errors have a `message` property and could also provide a `data` property for additional context.

shekharnwagh replied:
I moved the sync package dependency by moving the util to editor package. I went with using plain objects instead of Error object to keep it simple (no having create an instance and then attach additional fields). What would be other upside of using Error instances ?

chriszarate commented on packages/sync/src/types.ts:
```diff
@@ -52,20 +52,62 @@ export interface ProviderCreatorResult {
 	destroy: () => void;
 }
```
I know you followed my suggestion, but I see a implementation tweak that would probably be better. Instead of passing `onStateChange` to the provider, we can ask the result to expose an `on` method for attaching listeners. Since providers generally extend `Observable`, this should be an easy lift.

This may not be exactly right but points in the direction:

```suggestion
type ProviderOnEvent = ( eventType: 'status', ( state: SyncConnectionState ) => void ) => void;

// type overloads for future events can go here
// type ProviderOnEvent = ( eventType: 'other-event', ... ) => void

export interface ProviderCreatorResult {
	destroy: () => void;
	on: ProviderOnEvent;
}
```

Then the sync manager becomes responsible for attaching via `.on()` and calling the callback

shekharnwagh replied:
If I am understanding this, the provider would emit a 'status' event and manager listens to that and updates the store ? 

What are the advantages of this approach over passing callback ?

chriszarate replied:
Just centralized and final in API terms. If there is another event we want to listen to in the future, we don't need to make a change factory args to pass another callback. We can definitely do in a follow up PR if you want, but I think we should do it that way for the trunk PR.

shekharnwagh committed:
Add sync connection error handling

Creates a provider-agnostic system for sync providers to report
connection errors and display appropriate UI to users.

shekharnwagh committed:
Consolidate sync connection state into core-data

Removes the separate syncConnectionStore and moves connection state
management into core-data. Sync providers now receive an onStateChange
callback to report connection status for both entities and collections.

shekharnwagh committed:
Remove sync package dependency from editor

Moves sync error message handling from @wordpress/sync into editor
utils.

shekharnwagh committed:
Use event emitter pattern for provider status events

Replace callback injection with event listener pattern for reporting
provider connection status. Providers now expose an on() method that
allows attaching listeners after creation, rather than receiving
callbacks as constructor parameters.

alecgeatches reviewed (approved):
This looks good to me! I like that we're now storing entity status rather than ephemerally passing around signals when there is a connection issue. I'm excited to give this a whirl with versioning mismatches in https://github.com/WordPress/gutenberg/pull/74561 and prevent [reconnection issues](https://github.com/WordPress/gutenberg/pull/74075#issuecomment-3725787056) when those happen.

chriszarate commented on packages/sync/README.md:
```diff
@@ -50,6 +50,26 @@ Origin string for CRDT document changes originating from the local editor.
 
+### OnStateChangeCallback
+
+Undocumented declaration.
```
You might need to run the doc gen command, but these need to be populated.

shekharnwagh replied:
Added in 70092462d45

shekharnwagh committed:
Document sync provider connection state types

Adds JSDoc to OnStateChangeCallback, ProviderCreatorOptions,
SyncConnectionError, SyncConnectionState, and SyncConnectionStatus.
